### ISSUE TITLE
[google_sign_in_web] Migrate from Google Sign-In JavaScript Platform Library to Google Identity Services

### ISSUE THREAD

**Issue by jmagman** [flutter/flutter#88084]
Internal: b/233812803

Internal: b/259309242

Google Sign-In JavaScript Platform Library used by the google_sign_in web plugin [will be discontinued on March 31, 2023](https://developers.googleblog.com/2021/08/gsi-jsweb-deprecation.html).

It is being replaced by [Google Identity Services ](https://developers.google.com/identity/gsi/web)

https://github.com/flutter/plugins/blob/b548ae18646d1725a8e1f004c26bf42e4d362aa8/packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart#L24

### See also

* https://github.com/flutter/flutter/issues/101004

**Comment by jmagman** [comment]
\cc @ditman

**Comment by Levi-Lesches** [comment]
See also #82980

**Comment by ditman** [comment]
Thanks for reporting! I'll take a look at this ASAP; apologies for the nagging emails, I wasn't aware that this migration was going to happen or that we'd send out emails to eeeevery user of the plugin  :)

Speaking of the email, here's the full text we all received (included for SEO purposes on this issue :P)

> ## Google Developers
>
> ### _One or more of your web applications uses the legacy Google Sign-In JavaScript library. Please migrate your project(s) to the new Google Identity Services SDK before March 31, 2023._
>
> Hello Google Developer,
>
> As [announced](https://developers.googleblog.com/2021/08/gsi-jsweb-deprecation.html) earlier this month, the legacy [Google Sign-In web solution](https://developers.google.com/identity/sign-in/web/sign-in) will no longer be supported after **March 31, 2023**. You are receiving this message because one or more of your web applications uses the legacy Google Sign-In web solution.
>
> We’re writing to let you know that migration to the new [Google Identity Services](https://developers.google.com/identity/gsi/web) - our new family of Identity APIs that consolidate multiple identity offerings under one software development kit (SDK) - will be necessary to sign in users using their Google Account to your platform, after that date. **The SDK will use the new Sign In With Google client library.**
>
> #### What do I need to know?
> To help you find out which of your web apps or environments are affected, we’re sharing a list of your client ID(s) that use the legacy Google Sign-In web solution. This list should help you evaluate user impact and prioritize the migration work more efficiently:
>
> | App Name | client ID |
> |-----------|---------|
> | Your App Name | 00000000-y0urg00gl3s1gn1n4p1k3y.apps.googleusercontent.com |
>
> #### What do I need to do?
> At your earliest convenience, **migrate to the new Google Identity Services by following the [migration guide](https://developers.google.com/identity/gsi/web/guides/migration).**
>
> #### How can I find more information or share feedback?
> Be sure to read through the [Sign-In With Google resources](https://developers.google.com/identity/gsi/web/guides/overview) guide. If you have migration specific feedback to share, send an email to gis-migration-feedback@google.com.
>
>Sincerely,
>
> The Google Identity Services team


Luckily we have until 2023 to get this done ;P

Since google_sign_in is federated, we can start a new package for web that uses the new SDK. People will be able to test the new plugin while it's being developed, and once the new plugin is ready and tested, we can just change the endorsement to the new one.

I am not sure if we want to use this opportunity to rewrite the core plugin (a completely new platform interface, etc...). Any ideas @jmagman @blasten @stuartmorgan ?

**Comment by stuartmorgan** [comment]
> I am not sure if we want to use this opportunity to rewrite the core plugin (a completely new platform interface, etc...). Any ideas @jmagman @blasten @stuartmorgan ?

Is this using the new two-phase flow like the latest iOS SDK? If so, then we should seriously consider a set of breaking changes to all the interfaces across platforms to pick up that model. (We're waiting a bit to see what happens with the feedback that the new iOS SDK has received though in case anything changes as a result.)

**Comment by cybex-dev** [comment]
@ditman looking forward to this update, got a rough eta?

**Comment by Zujaj** [comment]
Any update regarding Flutter Web, because the [Google One-Tap Sign In](https://pub.dev/packages/google_one_tap_sign_in/example) package works only for Android.

**Comment by ditman** [comment]
@cybex-dev not yet, but I brought this up again to the team so it gets prioritized.

@Zujaj that package is developed by a third party and we don't support it. Find its repository [here](https://github.com/daewu14/google_one_tap_sign_in).

**Comment by kmeljko** [comment]
any news about this issue @ditman ?

**Comment by ditman** [comment]
No news, I'd personally would like to wrap this up by Q3 2022 so there's one Q for user testing/emergency fixes, and then we can release this at the end of Q4 2022 (right before the migration *must* happen)

**Comment by ditman** [comment]
See some interesting comments in the related ticket above:

> @brdaugherty commented [4 days ago](https://github.com/flutter/flutter/issues/101004#issuecomment-1091839454)
> ...
> Not mentioned in the above resources is the deprecation of the gapi.auth2 module found in the [Google API Client Library for JavaScript](https://github.com/google/google-api-javascript-client). This module is shared with Google Sign-in platform.js library. The deprecation affects both. It is very likely you've reviewed the developer migration guides already, but for completeness: for user [sign-in](https://developers.google.com/identity/gsi/web/guides/migration) and for [authorization](https://developers.google.com/identity/oauth2/web/guides/migration-to-gis#overview).

We'll need a new JS-interop layer to go with the plugin.

Also, since this seems to only affect web, we'll try to implement the already existing plugin API with the new JS API. [Docs](https://developers.google.com/identity/gsi/web/reference/js-reference). It seems that a migration to the new API with `ux_mode=popup` should be fairly straightforward.

**Comment by kevmoo** [comment]
@ditman – perhaps? Tha browser auth code there is pretty old. I've done some cleanup there, but it's pretty old.

In general, fewer Dart implementations of this logic would be better!

**Comment by ditman** [comment]
Immediate next steps to mitigate:

> Beginning April 30th, 2022 new applications must use the Google Identity Services library, existing apps may continue using the Platform Library until the deprecation date.

We'll do a small modification to the google_sign_in plugin suggested by @brdaugherty today so new applications can continue using the plugin after April 30th 2022.

**Comment by ditman** [comment]
> We'll do a small modification to the google_sign_in plugin suggested by @brdaugherty today so new applications can continue using the plugin after April 30th 2022.

This has been published as [`package:google_sign_in_web` v0.10.1](https://pub.dev/packages/google_sign_in_web/changelog#0101).

**Comment by maks** [comment]
@ditman I'm using `google_sign_in: ^5.4.0` in my Flutter web app and checking `pubspec.lock` I see:

```
 google_sign_in_web:
    dependency: transitive
    description:
      name: google_sign_in_web
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.10.2"
```
But when I run my app I still see on the Debug console:
```
Your client application uses libraries for user authentication or authorization that will soon be deprecated. See the [Migration Guide](https://developers.google.com/identity/gsi/web/guides/gis-migration) for more information.
```

**Comment by ditman** [comment]
@maks that is correct. `0.10.2` is still using the deprecated library. We're working on a fix with the GIS team. (We have until March 2023 to resolve this, but I hope to be done much earlier so people have plenty of time to adopt and test the new library out)

**Comment by maks** [comment]
Ah ok thanks @ditman , sorry I misunderstood: so the fix you mentioned previously to allow new applications to keep using the plugin after April is in 10.2 but it doesn't stop that deprecation message from showing.

**Comment by shatanikmahanty** [comment]
I created a new client ID today using firebase for flutter web. I am using latest google_signin package, but the error message shows up and I am unable to use google login. Is there any workaround?

**Comment by ramonvermeulen** [comment]
> I created a new client ID today using firebase for flutter web. I am using latest google_signin package, but the error message shows up and I am unable to use google login. Is there any workaround?

I don't think there is right now. We still need to wait for the major upgrade in the library to support the Google Identity Services API. Hopefully this will be released somewhere in the final quarter of 2022 so there is still time to migrate.

**Comment by ditman** [comment]
I'm actively working in this migration right now. There's a few kinks that need to be ironed out, though; the new SDK has some changes that affect how this integrates with your apps, for example:

> User sign-in status to their Google Account and your app are independent of each other, except during the sign-in moment itself when you know that the user has successfully authenticated and is signed into their Google Account. Users may remain signed-in, sign-out, or switch to a different Google Account while maintaining an active, signed-in session on your website.

These are the "migration guides" that I'm looking at to attempt to replicate the old behavior as much as possible to minimize changes:

* Authentication: https://developers.google.com/identity/gsi/web/guides/migration#object_migration_reference_for_user_sign-in (old vs new code)
* Authorization: https://developers.google.com/identity/oauth2/web/guides/migration-to-gis#library_quick_reference

Overall, I think this will make it much easier for Google Sign In to be used as an authentication/authorization provider by Firebase and others, but I'm not sure what's the impact in the case a user only used google authentication (and didn't manage their own users and sessions, those might need more work).

[Docs](https://developers.google.com/identity/gsi/web/guides/overview).

**Comment by ditman** [comment]
I am introducing a Dart JS-Interop layer package in the PR above so we can implement the new API. If you've read the documentation, you may be aware of some of the shortcomings of the new API, like:

* No token auto-renewal (tokens last 1h)
* The SDK no longer returns both ID and access_token at the same time, now users are supposed to give permissions to access their identity and THEN to access the Google APIs.

Anyway, I'll implement this as a new package (sibling of google_sign_in_web) to try to keep this API compatible, even though the new behavior is slightly different.

**Comment by Levi-Lesches** [comment]
> No token auto-renewal (tokens last 1h)

Does this mean users need to sign in again after 1hr of inactivity?

**Comment by ditman** [comment]
> Does this mean users need to sign in again after 1hr of inactivity?

@Levi-Lesches I don't think this has to do with inactivity, I think this means that Google will not maintain user sessions for your site anymore; you're supposed to use the google identity to create/link accounts to a user on your site. I think it's somewhat described in the docs [here](https://developers.google.com/identity/gsi/web/guides/overview#user_sign-in_to_your_site):

> You'll manage per user session state for sign-in to your site.
>
> User sign-in status to their Google Account and your app are independent of each other, except during the sign-in moment itself when you know that the user has successfully authenticated and is signed into their Google Account. Users may remain signed-in, sign-out, or switch to a different Google Account while maintaining an active, signed-in session on your website.

Also, with regards to Authorization tokens (oauth2 `access_token`, the ones that authorize access to `scopes`):

> access tokens have a short lifetime. If the access token expires prior to the end of the user's session, obtain a new token by calling `requestAccessToken()` from a user-driven event such as a button press.

The client-side SDK will not attempt to renew the token before it completely expires, it's up to your app to see that a request to an API failed, so a new token needs to be requested (from a user interaction). In order to have a renewable token, the `code` client needs to be used, where your server will keep the data fresh.

(The above is why I'm making this a separate package, the features are going to be slightly different than before, but since the old API still seems compatible (sign in/sign out/is signed in, and requestPermissions), I'm trying to create a package that "looks" like the old one, however some changes to our apps seem inevitable).

**Comment by cristhianA94** [comment]
Hi Ditman, thank you very much for the effort you are doing to migrate this library for web flutter, but could you enable the google_sign_in_web package you had uploaded, this was working for me to work with the Google Drive API I have in a web application in production for a client and is claiming me because it does not work.... I would be very grateful. Greetings!

**Comment by ditman** [comment]
@cristhianA94 we didn't disable anything with regards to google_sign_in_web, in fact, it should contain an exemption to continue working as intended. Please, ensure that the version your app is using is newer than `0.10.1`.

See the [changelog](https://pub.dev/packages/google_sign_in_web/changelog).

**Comment by cristhianA94** [comment]
Sorry Ditman, you are right, it still works, I think that as the package is listed as "Unlist" in pub.dev, when I installed another dependency it was removed. I already tested it and everything is ok, thank you very much.

**Comment by stuartmorgan** [comment]
> I think that as the package is listed as "Unlist" in pub.dev, when I installed another dependency it was removed.

Unlisted packages just don't show up in pub.dev search results; it has no effect on how they are used, or on the resolver. Also, if you depend on a recent version of `google_sign_in`, `google_sign_in_web` can't be removed by `pub`, since [it's a dependency of the package](https://github.com/flutter/plugins/blob/88dc5e33a327787a89b0e014425e8f2c528cfe10/packages/google_sign_in/google_sign_in/pubspec.yaml#L29).

**Comment by ditman** [comment]
Hey all!

I just pushed a quick and dirty early draft (no tests! probably doesn't work yet, **I haven't attempted to integrate it yet!**) but so people can see what this _migration_ is going to be about, check it out here:

* https://github.com/ditman/plugins/tree/google-sign-in-gis-web/packages/google_sign_in/google_sign_in_gis_web

Check it out and let me know what you think! It has a README, and it uses a separate package to interop with the new JS SDK (which you can also use now!)

_PS:  The path dependency in the plugin is pending to land here: https://github.com/flutter/packages/pull/2688_

_(PPS: I'm at an all-hands team meeting, so my ability to respond this week is limited. I just wanted to offer a sneak peek of the work that's happening here, and also "preserve" the code somewhere in github, rather than only in my machine 😅)_

**Comment by ditman** [comment]
I did some more fiddling with the new SDK last friday, trying to keep the plugin API compatible, **unsuccessfully.** You can see the experiments in the plugin [here](https://github.com/ditman/plugins/commit/8d3f5d9a9b2e7e5147d4ae64a78dab3a3311289f), and attempted to wire it to the [example app](https://github.com/ditman/plugins/commit/ab385774ff52755c8e5fa5e59ccdf3eb02d8e007) for validation.

My approach was:

* Use the one-tap prompt for the implementation of silentSignIn; it's not totally *silent*, but at least we don't have to do crazy stuff to maintain the "last seen credential" around. It also felt more consistent with the spirit of the new SDK.
* Use the oauth2 implicit flow to have "authentication + authorization" in one popup (which the "id" library doesn't provide); this opens a very similar popup as with the current version of `google_sign_in_web`, but it will **NOT return the Authentication of the user (`credential`), only the Authorization (`access_token`).**
    * I attempted to "fill in" the user information by requesting a couple of extra "userinfo" scopes (`userinfo.email`, `userinfo.profile`) for the People API (see the file `network.dart`), and attempting to request info from the [people.get](https://developers.google.com/people/api/rest/v1/people/get) method.
        * Adding the extra `userinfo` scopes (or the `email` scope that was in the app) caused the Sign-In popup to display an error: **`invalid_scopes`**. All scopes I passed in were marked as `invalid`, and no further information was given _(maybe because my test app is not [verified](https://developers.google.com/identity/protocols/oauth2/production-readiness/brand-verification)?)_

The error above uncovered a small issue (that AFAIK has already been reported); the oauth2 popup does *not* notify the application of a wrong/unsucessful/aborted login attempt, so if **the requestToken process doesn't finish through the happy path, our plugin keeps `await`ing for a Future that will never be rejected.**

At this point, and barring programmer error (that's why I'm linking to the diffs too 😜), I'm not entirely sure that we can have a version of `google_sign_in` for the web that is compatible with the old API, so **unless the new SDK changes, it's likely that we'll need a new package for the web, incompatible with the current `google_sign_in_platform_interface`.**

### Next steps:

* The js-interop layer is not complete, it's not exposing any of the [renderButton](https://developers.google.com/identity/gsi/web/reference/js-reference#google.accounts.id.renderButton) API or its configuration attributes. **Add this**.
* Build a Flutter widget that adapts to the size of its (web) contents, so we can render the button and have it resize while reflowing the Flutter app around it.
* Then, create a new plugin that is a very thin layer on top of the SDK, just to add some usability to it, for example:
    * Mimic the SDK API to follow their flows: `id.renderButton`, `id.prompt` (for one-tap) or `oauth2.requestToken`.
    * Expose Authentication and Authorization events as Streams, so the UI can be built in a "rx" fashion, rather than being Future-based (which can lead to the UI of the app hanging). (It is easy to go from Streams to Futures too, if anybody wants those.)
    * ...
    * Profit?

**Meanwhile:**

* Discuss the feasibility to keep *a single plugin* for all platforms with the current "state of the art" of the available GIS SDKs.

----

Anyway, thanks for reading this wall of text. If anybody comes up with a different set of solutions, feel free to share. Otherwise, I'll start modifying the js-interop layer and maybe adding the "Elastic" Web widget :)

**Comment by xster** [comment]
Changing to P3 to match b/233812803

**Comment by hpomgithub** [comment]
Hi @ditman ,
After the completion of migration process do we need to forcefully upgrade our present billing scheme of FirebaseAuth to the new 'Auth with identity platform' in our firebase console, in order to use the new plugins made for google sign in using new identity services api ?

**Comment by ditman** [comment]
> After the completion of migration process do we need to forcefully upgrade our present billing scheme of FirebaseAuth to the new 'Auth with identity platform' in our firebase console, in order to use the new plugins made for google sign in using new identity services api ?

@hpomgithub I don't know, I think the migration should be transparent in that regard, since all the code knows about your firebase account is the "client ID" used for Google Auth itself, but **you should ask that through Firebase customer support to be 100% certain.**

(The fix I'm working on is completely independent of Firebase inner workings, this plugin can be used directly by projects using the Google Cloud console directly, without any Firebase whatsoever)

**Comment by hpomgithub** [comment]
Hi @ditman 
Thank You, as I got my answer after checking with firebase team. One thing I I want to ask is about the new plugin which you are developing (to sign_in with  google) , can we use it in a same way to carry out authentication using firebase 'FirebaseAuth.instance.signInWithPopup(GoogleAuthProvider())', like we use to do earlier because firebase security depends majorly on the auth part.

**Comment by ditman** [comment]
To everybody interested in this, **please do check out the PR and Demo** app that I have just posted, here:

* https://github.com/flutter/plugins/pull/6921

I still need to rewrite most tests (the JS mocks I used to use are not valid anymore), but you can take a look. Particularly at the subtle changes in behavior of `silentSignIn` and `signIn`.

**Please, do provide feedback!** To me, the plugin "works" (but I don't know all possible use cases, so your definition of "works" and mine might vary!)

**Comment by ditman** [comment]
The PR has been updated with a version of the branch that should be much easier to use directly from Git (no path dependencies anymore), with a dependency override pointing [here](https://github.com/ditman/plugins/tree/google-sign-in-web-using-gis/packages/google_sign_in/google_sign_in_web):

```
  google_sign_in_web:
    git:
      url: https://github.com/ditman/plugins.git
      ref: google-sign-in-web-using-gis
      path: packages/google_sign_in/google_sign_in_web
```

**Comment by hpomgithub** [comment]
Hi @ditman ,
From where we can the document to implement the new one and use with firebase auth like we used to do it with previous google_sign_in_web.

**Comment by avatarKaran** [comment]
> The PR has been updated with a version of the branch that should be much easier to use directly from Git (no path dependencies anymore), with a dependency override pointing [here](https://github.com/ditman/plugins/tree/google-sign-in-web-using-gis/packages/google_sign_in/google_sign_in_web):
> 
> ```
>   google_sign_in_web:
>     git:
>       url: https://github.com/ditman/plugins.git
>       ref: google-sign-in-web-using-gis
>       path: packages/google_sign_in/google_sign_in_web
> ```

hey, I tried adding that to the pubspec.yaml and i got:
Because google_sign_in 5.4.3 depends on google_sign_in_web ^0.10.0 and no versions of google_sign_in match >5.4.3 <6.0.0, google_sign_in ^5.4.3 requires google_sign_in_web from hosted.
 because app depends on both google_sign_in_web from git and google_sign_in ^5.4.3, version solving failed.)

**Comment by stuartmorgan** [comment]
That sounds like you added it in `dependencies` rather than `dependency_overrides`; only the latter will work.

**Comment by avatarKaran** [comment]
> That sounds like you added it in `dependencies` rather than `dependency_overrides`; only the latter will work.

thank you. that solved it. is it possible you could tell me the approach to this new method?

patient opens flutter web app>
2) clicks google sign-in button:
3)google identity services pops  open upon click depending on singinSilently(). thats if the browser detects a session.
ELSE itll open up the Oauth. which will then return a googleaccount via:  GoogleSignInAccount? googleaccount = await _googleSignIn.signIn();

for some reason my app says: The OAuth token was not passed to gapi.client, since the gapi.client library is not loaded in your page.

**Comment by hpomgithub** [comment]
Hi,
From where we can get the document to implement the new implementation and use with firebase auth, like we used to do with previous google_sign_in_web ?

**Comment by ditman** [comment]
> From where we can get the document to implement the new implementation and use with firebase auth, like we used to do with previous google_sign_in_web.

@hpomgithub You can install the version of the plugin in the PR by adding a `dependency_overrides`, as described [here](https://github.com/flutter/flutter/issues/88084#issuecomment-1378143035).

With regards to `firebase_auth`, I am not 100% positive how to use this. We should bring somebody from the firebase repo (I'll reach out to them, and point them to this issue).

> That sounds like you added it in `dependencies` rather than `dependency_overrides`; only the latter will work.

@stuartmorgan this is a good point, I updated the "setup" instructions in my comment above, so it's clearer.

> thank you. that solved it. is it possible you could tell me the approach to this new method?
>
> patient opens flutter web app>
> 2) clicks google sign-in button:
> 3) google identity services pops open upon click depending on singinSilently(). thats if the browser detects a session. but since it always returns Null. how do you obtain the access token via signinsilently().
> ELSE itll open up the Oauth. which will then return a googleaccount via: GoogleSignInAccount? googleaccount = await _googleSignIn.signIn();

@avatarKaran thanks for giving the new plugin a shot, as you can read in the description of the PR [here](https://github.com/flutter/plugins/pull/6921), the small differences that you're finding are the reason why the new version is (most probably) be released as a MAJOR version, even though there shouldn't be many changes to do in your web app.

The root of the problem is that:

* The new Authentication API (idToken) does NOT return `accesToken`s anymore, and
* The new Authorization API (oauth2, accessToken) does NOT return `idToken`s anymore.

In order to try to minimize those two differences of behavior, we do the following:

* `signInSilently()` always returns `null`, because the new SDK does **not** authorize the scopes on sign in. If the user authenticates with the One Tap card, we keep the `idToken` for later, but we tell the plugin that the authentication was not successful (because we don't have the `accessToken` yet).
  * (In future versions of the web package, as people migrate to the new world of expiring Authorization credentials, we can start returning the object we receive in silentSignIn, instead of always null, but not in the first version!)
* `signIn()` behaves in two ways:
  * If the user has accepted the `signInSilently()`, we call the oauth2 authorization process (the popup), and if successful, return a User Information object with the `idToken` from the previous signInSilently call. We store the `accessToken`.
  * If the user has NOT accepted the `signInSilently()`, we add a couple more scopes to the oauth2 authorization process, so we can do a People Request API to retrieve as much User Information as we can immediately after a successful authorization (unfortunately, `idToken` will be null in this case).

The way that the new API wants to be used is the following:

* Completely separate sign in (authentication) from the oauth2 popup (authorization).
* Users sign in with the signIn and signInSilently methods. These two do *not* handle `scope`.
* Then, users get permissions with the `requestScopes` method.
  * The oauth2 authorization is now very short lived (1h, I think), and will **not auto-renew** as before.
    * Your app *must* keep track of failed requests to APIs, and re-authorize the user as needed.

See the official documentation of the JS migrations here: 

* Authentication: https://developers.google.com/identity/gsi/web/guides/migration
* Authorization: https://developers.google.com/identity/oauth2/web/guides/migration-to-gis

> for some reason my app says: The OAuth token was not passed to gapi.client, since the gapi.client library is not loaded in your page.

@avatarKaran this is expected (but in my latest changes I've made this only show up in debug mode). The google_sign_in_web plugin does NOT use the `gapi` library anymore, so we don't even download its JS. The new SDK warns that `gapi` is not in the page to interact with it, but this is **only a warning**, and shouldn't affect functionality. (The warning now only shows up in debug mode)

**Comment by ditman** [comment]
@hpomgithub: I've confirmed with the [Flutterfire](https://github.com/firebase/flutterfire) maintainers (thanks @Lyokone!) that `firebase_auth` works fine with the new version of the plugin. (🥳)

**Comment by ditman** [comment]
I'll have a meeting with members of the GIS team to discuss the "tricks" that this new implementation does to stay API compatible with the old one, and discuss a roadmap with the upcoming changes to align the plugin with what the new SDK wants to do (primarily: separating authentication from authorization).

I'll share a document with the roadmap for the upcoming changes ASAP (hopefully next week, I need to cordinate a bunch of people to attend the same meeting :))

**Comment by hpomgithub** [comment]
> > With regards to `firebase_auth`, I am not 100% positive how to use this. We should bring somebody from the firebase repo (I'll reach out to them, and point them to this issue).
> 
> @hpomgithub: I've confirmed with the [Flutterfire](https://github.com/firebase/flutterfire) maintainers (thanks @Lyokone!) that `firebase_auth` works fine with the new version of the plugin. (🥳)

Thank you @ditman for clarifying this. Can you mention any document link explaining the implementation of the new Api and carry out the authentication using firebase auth.

**Comment by ditman** [comment]
> hopefully next week, I need to cordinate a bunch of people from different teams to attend the same meeting

The meeting will happen next Tuesday, Jan 24th at 3PM (UTC -08:00). I'll try to share a public doc later that day, or on Wednesday, Jan 25th.

**Comment by hpomgithub** [comment]
> > hopefully next week, I need to cordinate a bunch of people from different teams to attend the same meeting
> 
> The meeting will happen next Tuesday, Jan 24th at 3PM (UTC -08:00). I'll try to share a public doc later that day, or on Wednesday, Jan 25th.

Thank you @ditman

**Comment by mikron123** [comment]
Does anyone has a web code snippet about the new syntax combined with Firebase Auth?

**Comment by avatarKaran** [comment]
> I'll have a meeting with members of the GIS team to discuss the "tricks" that this new implementation does to stay API compatible with the old one, and discuss a roadmap with the upcoming changes to align the plugin with what the new SDK wants to do (primarily: separating authentication from authorization).
> 
> I'll share a document with the roadmap for the upcoming changes ASAP (hopefully next week, I need to cordinate a bunch of people from different teams to attend the same meeting :))

thank you for the quick responses. youre doing great. will look forward to your updates. my app is a flutter web app with a django back end.

**Comment by edu-correia** [comment]
Hello, everyone!
I'm giving a try on this new version and for me it's working fine, but my only problem is that I need to have access on the `idToken` to send it to my backend server, but I can't get it anymore, only the `accessToken`. However, looking at the logs, I can see that the library prints out exactly what I need, the `idToken`/`credential`.


`[GSI_LOGGER]: Message received: {"type":"response","response":{"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx","client_id":"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx","credential":"eyJhbGciOiJSUzI1NiIsImtpZCI6ImFmYzRmInE2NTk5ZmY1ZjYzYjcyZGM1MjI0MjgyNzg2ODJmM2E3ZjEiLCJ0eXAiOiJKV1QifQ.eyJpc3MiOiJodHRwczovL2FjY291bnRzLmdvb2dsZS5jb20iLCJuYmYiOjE2NzQ1NzE3NjcsImF1ZCI6IjkzMjkxNjkwMDk1MS12cjdnbHJ1cDZlcm1kdW3obnFzNDgyOXNtZDJtZWgzaC5hcHBzLmdvb2dsZXVzZXJjb250ZW50LmNvbSIsInN1YiI6IjEwNjE5NzA2NjYyMjUyODU3OTc0MSIsImhkIjoibW90b3JvbGEuY29tIiwiZW1haWwiOiJlaHNhbnRvc0Btb3Rvcm9sYS5jb20iLCJlbWFpbF92ZXJpZmllZCI6dHJ1ZSwiYXp0IjoiOTMyOTE2OTAwOTUxLXZyN2dscnVwNmVybWR1ZGhucXM0ODI5c21kMm1laDNoLmFwcHMuZ29vZ2xldXNlcmNvbnRlbnQuY29tIiwibmFtZSI6IkVkdWFyZG8gSGVucmlxdWUgZG9zIFNhbnRvcyBDb3JyZWlhIiwicGljdHVyZSI6Imh0dHBzOi8vbGgzLmdvb2dsZXVzZXJjb250ZW50LmNvbS9hL0FFZEZUcDRlRW04WExaWDg5OXE3ZVNUTC1aZVZkdHRqNE9GTEx2WDdJcTlSPXM5Ni1jIiwiZ2l2ZW5fbmFtZSI6IkVkdWFyZl8iLCJmYW1pbHlfbmFtZSI6IkhlbnJpcXVlIGRvcyBTYW50b3MgQ29ycmVpYSIsImlhdCI6MTY3NDU3MjA2NywiZXhwIjoxNjc0NTc1NjY3LCJqdGkiOiJhOTQ2NDA0NTkyNjllNTkyNDExMzMyMTg3NWNkOTRlMDViNGRiOTgxIn0.gNSA96rQJIc-BPkVun9BDLE8dsCdZjudkvxDqpvNHUjCR-W0vlQywVdlKW3zcPIP-wRkNTyBxjTPw5P8C4hGW8pOmzlBI_l2zvD8k247z1AziJfcGG6hd5jS_GDR251s6xDeab_6EI4bQXMybS9Sr-CS5w6x3GmC0TmdIFFMpYWN4hCDMl_RLecOcS6zJLcBhR2CIFoLm0TVijQuP5u27IDIV0e1NojGS4Rf96vXQ-8K-2qm-2SB9N0PDzjWH1v2m6C9-Cgln0OpQhSpoLbrBKTkzxF5faCg6mnFqBKzPAk0MbSeDpaz5mQbl9V7Wyb8t9ddC6cceNoAd80Dhugdjg","select_by":"auto"},"announcement":"Signed in as XXXXXXXXX"}`

So, what method can I call to get this `credential` value?

**Comment by avinash-bsharp** [comment]
Does anyone has a working example of google sign in on flutter web with gis and without Firebase Auth?

**Comment by kevmoo** [comment]
@avinash-bsharp tried https://pub.dev/packages/google_sign_in ?

**Comment by avinash-bsharp** [comment]
@kevmoo I need it for web. This package has the old documentation I think https://pub.dev/packages/google_sign_in_web. I am getting mails for deprecation.

**Comment by kevmoo** [comment]
@ditman should be working on rolling this forward to the latest APIs.

**Comment by jorgeroncero** [comment]
I can confirm that the PR from @ditman is working without breaking anything in our app. **BUT make sure that you enable People Google API for your project or the scope request will fail.**

@ditman the People API request is asking for photos, names and emailAddresses scopes even if I ask only for email, is that the intended behaviour?

For those asking, we are using Google Sign In method and Firebase Auth this way.

```dart
final _googleSignIn = GoogleSignIn(
        scopes: [
          'email',
        ],
        clientId: DevicePlatform.isWeb
            ? const String.fromEnvironment('WEB_GOOGLE_OAUTH_CLIENT_ID')
            : null,
      );

final googleSignInAccount = await _googleSignIn.signIn();

final googleAuth = await googleSignInAccount.authentication;

final googleAuthCredential = GoogleAuthProvider.credential(
        accessToken: googleAccessTokenEntity.accessToken,
        idToken: googleAccessTokenEntity.idToken, //Should be null now in WEB
      );

await firebaseAuth.signInWithCredential(googleAuthCredential);
```

**Comment by Prajun2010** [comment]
Hello @jorgeroncero @ditman,
In our project, we have used google_sign_in and firebase for authentication in both mobile and web platforms.
Recently we've received the migration notice. So, to fix that we've been trying to implement all the possible ways which you have described in this portal.

https://github.com/flutter/flutter/issues/88084#issuecomment-1378143035

But we aren't able to resolve it, and every time we try to override the dependencies, we are getting this message:
"Package not available (the pubspec for google_sign_in_web 0.11.0 from git has version 0.10.2+1)"



Is there anything I've missed over here, or implemented in the wrong way?

**Comment by jorgeroncero** [comment]
@Prajun2010 can you show your pubspec.yaml? Or at least your google_sign_in dependency and the google_sign_in_web dependency_override

**Comment by Prajun2010** [comment]
@jmagman 
![image](https://user-images.githubusercontent.com/42827734/215084661-e1697ab0-3a1c-4246-acae-b794a2db2b20.png)

**Comment by jorgeroncero** [comment]
The pubspec seems right. 

The error could be caused because the previous version installed in your cache was 0.10.2+1. Have you tried to do a `flutter clean` and then `flutter pub get`? The `pubspec.lock` file must be forcing to use the previous version, not sure why.

**Comment by hpomgithub** [comment]
Hi @ditman 
Big ups for the work you did! Just tried out the new version. It is working perfectly fine when we use 'GoogleSignIn().signIn()' (after enabling people's api) - with firebase auth,
But it is unable to get the credential when we use GoogleSignIn().signInSilently(). It will be great if you can suggest steps to get the credential obtained while using signInSilently().
Also do we need to sign in every hour as access token expiration is now 1 hour?

**Comment by ditman** [comment]
> So, what method can I call to get this credential value?

@edu-correia you should be able to get both the `idToken` (authentication credential) and `accessToken` (authorization token) (if the user has given you permission to have them) with the [authentication](https://pub.dev/documentation/google_sign_in/latest/google_sign_in/GoogleSignInAccount/authentication.html) getter of the `GoogleSignInAccount` after `signIn`. Note that in the new implementation, the `idToken` may be `null`, if the user doesn't go through (or cancels) the `signInSilently` (OneTap UX) flow, in which case the plugin will fall-back to a People API request to retrieve the basic profile information of the user (name, email, photo and google User ID). You will have everything you need about the user (except its JWT representation as `idToken`).

> Does anyone has a working example of google sign in on flutter web with GIS and without Firebase Auth?

@avinash-bsharp yes, check the [`package:google_sign_in` example app](https://github.com/flutter/plugins/tree/main/packages/google_sign_in/google_sign_in/example), which I have deployed for testing [here](https://dit-gis-test.web.app) with my upcoming changes as a dependency override.

> BUT make sure that you enable People Google API for your project or the scope request will fail.

@jorgeroncero this is a great point that I had totally forgotten to document. Back to the PR!

> the People API request is asking for photos, names and emailAddresses scopes even if I ask only for email, is that the intended behaviour?

@jorgeroncero this is correct, I'm attempting to return the same data that the previous version of the SDK gave you after Sign In. The People API request will only happen if the `signInSilently` doesn't authenticate the current user (or the user aborts the One Tap UX flow). In that case, the plugin attempts to retrieve the information from the People API after the oauth popup is done, by doing an extra REST request. If `signInSilently` is successful, the People API request won't happen.

> It is unable to get the credential when we use GoogleSignIn().signInSilently(). It will be great if you can suggest steps to get the credential obtained while using signInSilently().

@hpomgithub Unfortunately, I had to do this for the new version to be "backwards compatible" with the current implementations. Since the old `silentSignIn` did both Authentication and Authorization, it could return a full `GoogleSignInAccount` with a complete `authentication` (that contained both `idToken` and `accessToken`).

In the new SDK, however, the One Tap UX flow only returns the `idToken`, which is **half** of the `authentication`. Then on `signIn` we retrieve the other half: `accessToken` (with the authorized `scopes`). So only after the user has done `signIn` the plugin can say that the `authentication` is complete.

TL;DR: **we simulate that `signInSilently` fails on the web, so current apps are "forced" to call `signIn`, and thus obtain a complete `GoogleSignInAccount`.**

> Also do we need to sign in every hour as access token expiration is now 1 hour?

@hpomgithub I was also surprised to learn this, but... yes!

> ### Token expiration
>
> By design, access tokens have a short lifetime. If the access token expires prior to the end of the user's session, obtain a new token by calling requestAccessToken() from a user-driven event such as a button press.

_From the [Docs](https://developers.google.com/identity/oauth2/web/guides/use-token-model#token_expiration)._

For longer lived sessions, the new SDK proposes that you use the "Code" client, which means: you talk to the Google services from your backend, on behalf of the user. [Docs](https://developers.google.com/identity/oauth2/web/guides/use-code-model).

**Comment by ditman** [comment]
**Question for all: Should the new web plugin return an "authenticated-but-not-authorized" `GoogleSignInAccount` if `signInSilently` is successful (requires apps to call `signIn` or, much better, [`requestScopes`](https://pub.dev/documentation/google_sign_in/latest/google_sign_in/GoogleSignIn/requestScopes.html) before being able to perform REST requests), or should it only return a "complete" `GoogleSignInAccount` once the `accessToken` is granted?**

The new SDK is going in the direction of completely separating Authentication from Authorization, so from a "Google Identity Services", separating "isSignedIn" from "isAuthorized" is better, and since we're writing a _Migration Guide_, we might as well add that extra bit of backwards incompatibility :)

To minimize code changes, however, it is better that `signInSilently` "fails", as it does now (that enabled the example app to work without any significant changes!)

**Comment by ditman** [comment]
Please, do take a look at the "Migration guide"/"FAQ" on the new README, and do comment in the PR if you feel like there's missing stuff, or anything needs more clarification:

### [New README.md](https://github.com/ditman/plugins/blob/google-sign-in-web-using-gis/packages/google_sign_in/google_sign_in_web/README.md)

**And thank you everyone for your time testing this package!**

**Comment by hpomgithub** [comment]
@ditman Thanks for clarifying everything in detail. 


> @hpomgithub I was also surprised to learn this, but... yes!
> 
> >

Maybe Firebase Auth can help in renewing the tokens automatically as I have signed in using firebase yesterday with the new version & it was still signed in even today. What you think about this?

**Comment by ditman** [comment]
> Maybe Firebase Auth can help in renewing the tokens automatically as I have signed in using firebase yesterday with the new version & it was still signed in even today. What you think about this?

@hpomgithub for user authentication (login) yes, I even "recommend" it on the migration guide. For user **authorization** (give me access to your Contacts), I don't think it'll help.

**Comment by Ianpot12** [comment]
@ditman I do the request but the Id Token returns null, how can I fix this ~ Flutter Web? Thanks in advance .

**Comment by ditman** [comment]
> I do the request but the Id Token returns null, how can I fix this ~ Flutter Web? Thanks in advance.

@Ianpot12 I've documented that behavior in the README already, here:

* https://github.com/ditman/plugins/blob/google-sign-in-web-using-gis/packages/google_sign_in/google_sign_in_web/README.md#idtoken-is-null-in-the-googlesigninaccount-object-after-signin

What info do you need from the `idToken`?

**Comment by Ianpot12** [comment]
@ditman I want to send it our backend server

**Comment by ditman** [comment]
@Ianpot12 @edu-correia for you to get an idToken with the new plugin, you need to use `signInSilently`, which currently is the only flow that computes and returns a valid idToken. If users do *not* complete the sign in successfully, the idToken will not be produced (because I cannot fake Google server private keys from the plugin).

**Comment by Ianpot12** [comment]
@ditman Thanks !! , This is my method, how do I get the idToken from 
signInSilently() call if it was successful? Thanks for the help!!

<img width="681" alt="Screenshot 2023-02-02 at 20 47 56" src="https://user-images.githubusercontent.com/47270662/216421196-6e0d5950-641d-4a13-90b2-a8fa80a48294.png">

**Comment by ditman** [comment]
`_googleSignIn.signInSilently()` might never return a user, you need to use both `signInSilently` and `signIn`.

After `signIn`, if the user has authorized the OneTap flow to authenticate them, you'll get an `idToken` in the `GoogleSignInAccount.authentication` getter [(this one)](https://github.com/flutter/plugins/blob/main/packages/google_sign_in/google_sign_in/lib/google_sign_in.dart#L91).

**Comment by Ianpot12** [comment]
@ditman sorry man, but something likes this the IdToken still returns null
<img width="890" alt="Screenshot 2023-02-02 at 21 22 48" src="https://user-images.githubusercontent.com/47270662/216429778-840365ab-c64c-4e88-85e7-cf7dfa65680b.png">

**Comment by ditman** [comment]
@Ianpot12 not exactly like that (but close). `signIn` needs to be triggered from a user interaction (like a button press), so the browser won't block the auth pop-up.

Check the example app, it's very similar to yours (except it uses regular setState to propagate users from the currentUserChanged). Source:

* on initState, it [sets the onCurrentUserChanged listener, and calls signInSilently](https://github.com/flutter/plugins/blob/main/packages/google_sign_in/google_sign_in/example/lib/main.dart#L54)
* [signIn gets called](https://github.com/flutter/plugins/blob/main/packages/google_sign_in/google_sign_in/example/lib/main.dart#L106) on a [button click](https://github.com/flutter/plugins/blob/main/packages/google_sign_in/google_sign_in/example/lib/main.dart#L146-L149)

**Comment by Ianpot12** [comment]
@ditman  Thank you so much, I was very close to give up !
<img width="890" alt="Screenshot 2023-02-02 at 22 02 23" src="https://user-images.githubusercontent.com/47270662/216437399-8094980f-5983-4f8a-89ea-54a23a1ae601.png">
!!

**Comment by ditman** [comment]
@Ianpot12 do you mind pasting your working example as text, so people can find it when searching github by "_googleSignIn.signIn()" or similar? Thanks!!

You can use "triple backtick syntax" and github will do some syntax coloring even!

<pre>
```dart
// your dart code
```
</pre>

**Comment by Ianpot12** [comment]
Flutter / Dart gis migration with googleSignIn() 


My  dependencies
``` 
  google_sign_in: ^5.4.4
  google_sign_in_web: ^0.10.0

dependency_overrides:
  google_sign_in_web:
    git:
      url: https://github.com/ditman/plugins.git
      ref: google-sign-in-web-using-gis
      path: packages/google_sign_in/google_sign_in_web

```

Code ~ BLOC State management


```
Call this method first on page Load

silentUserLogin() async {
    _googleSignIn.onCurrentUserChanged.listen((GoogleSignInAccount? account) {
      emit(state.copyWith(currentUser: account));

      if (state.currentUser != null) {
        handleGetContact(state.currentUser!);
      }

    });

    _googleSignIn.signInSilently();
  }

```
```
Then this method onTap() ~ Login button

  login(Function onSuccess) async {
    try {
      await _googleSignIn.signIn();
    } catch (error) {
      print(error);
    }
    emit(state.copyWith(currentUser: _googleSignIn.currentUser));

    if (state.currentUser != null) {
      state.currentUser!.authentication.then((GoogleSignInAuthentication value) {
       print(value.idToken!)
      });
    }
  }

```

**Comment by Prajun2010** [comment]
> The pubspec seems right.
> 
> The error could be caused because the previous version installed in your cache was 0.10.2+1. Have you tried to do a `flutter clean` and then `flutter pub get`? The `pubspec.lock` file must be forcing to use the previous version, not sure why.

I've done that too, but the error is persistent.

**Comment by ditman** [comment]
@Prajun2010 I'm waiting for one more review, then land the PR. That'll create a new major version of the package in pub.dev, so you should be able to set a more "normal" override, before you fully commit to upgrading the package.

(Note that `package:google_sign_in` will also have a major version release, so people can control (and test) when they get the newest code for the web)

**Comment by ramonvermeulen** [comment]
For some reason `onCurrentUserChanged` doesn't get fired when I login through `signInSilently()`, using the following code:

```
    _googleSignIn.onCurrentUserChanged.listen((GoogleSignInAccount? account) {
      print("user changed");
    });
    _googleSignIn.signInSilently();
```

Any idea how I can debug this?

**Comment by ditman** [comment]
> For some reason onCurrentUserChanged doesn't get fired when I login through signInSilently(), using the following code:

@ramonvermeulen This is (sadly) expected, let me know if this bit of documentation explains this situation successfully:

https://github.com/ditman/flutter-plugins/tree/google-sign-in-web-using-gis/packages/google_sign_in/google_sign_in_web#signinsilently-always-returns-null

**Comment by ditman** [comment]
A change that I can think of is to return a user object IF there were no initial `scopes` requested. I think that'd be a fair compromise between the old and the new behavior.

**Comment by ramonvermeulen** [comment]
> > For some reason onCurrentUserChanged doesn't get fired when I login through signInSilently(), using the following code:
> 
> @ramonvermeulen This is (sadly) expected, let me know if this bit of documentation explains this situation successfully:
> 
> https://github.com/ditman/flutter-plugins/tree/google-sign-in-web-using-gis/packages/google_sign_in/google_sign_in_web#signinsilently-always-returns-null

@ditman Hmm I see, I'm still struggling a bit finding a correct way to migrate this implementation.

In the old implementation I listen on `_googleSignIn.onCurrentUserChanged` to change the "authenticated" state of the user within the application. On page load a `signInSilently()` got triggered to sign-in the user in the background if he already signed-in manually before. And on the login button the `signIn()` will get triggered manually. Both of these scenario's used to fire an event in the `onCurrentUserChanged` listener, but currently only `signIn` does that.

If I understand the explanation correctly I now have to request scopes after the `signInSilently` to trigger the `onCurrentUserChanged` event listener. (Because authentication and authorization is now separated)

For example if I try the following:

```
silentUserLogin() async {
  _googleSignIn.onCurrentUserChanged.listen((GoogleSignInAccount? account) {
    print("user changed"); // never gets called
  });
  await _googleSignIn.signInSilently();
  await _googleSignIn.requestScopes(_googleSignIn.scopes);
}
```

I get this output:

```
[GSI_LOGGER-TOKEN_CLIENT]: Instantiated.
[GSI_LOGGER]: Setup message received: {"type":"readyForConnect","channelId":"xxx"}
[GSI_LOGGER]: Setup message received: {"type":"readyForConnect","channelId":"xxx"}
[GSI_LOGGER]: Message received: {"type":"command","command":"resize","height":360}
[GSI_LOGGER]: Message received: {"type":"command","command":"cancel_protect_start"}
[GSI_LOGGER]: Message received: {"type":"command","command":"resize","height":147}
[GSI_LOGGER]: Message received: {"type":"command","command":"resize","height":147}
[GSI_LOGGER]: Message received: {"type":"command","command":"start_auto_select"}
[GSI_LOGGER]: Message received: {"type":"command","command":"cancel_protect_start"}
[GSI_LOGGER]: Message received: {"type":"response","response":{"clientId":"xxx","client_id":"xxx","credential":"xxx","select_by":"auto"},"announcement":"Ingelogd als r.vermeulen@xxx.com"}
[GSI_LOGGER]: Response received: {"clientId":"xxx","client_id":"xxx","credential":"xxx","select_by":"auto"}
[GSI_LOGGER-TOKEN_CLIENT]: Starting popup flow.
[GSI_LOGGER-OAUTH2_CLIENT]: Starting popup timer.
[GSI_LOGGER-OAUTH2_CLIENT]: Checking popup closed.
[GSI_LOGGER-OAUTH2_CLIENT]: Checking popup closed.
[GSI_LOGGER-OAUTH2_CLIENT]: Checking popup closed.
[GSI_LOGGER-OAUTH2_CLIENT]: Checking popup closed.
[GSI_LOGGER-TOKEN_CLIENT]: Handling response. {"access_token":"xxx","token_type":"Bearer","expires_in":3599,"scope":"email profile https://www.googleapis.com/auth/userinfo.profile openid https://www.googleapis.com/auth/admin.directory.user.readonly https://www.googleapis.com/auth/userinfo.email https://www.googleapis.com/auth/cloud_search.query","authuser":"0","hd":"xxx.com","prompt":"none"}
[GSI_LOGGER-OAUTH2_CLIENT]: Popup timer stopped.
[GSI_LOGGER-TOKEN_CLIENT]: Trying to set gapi client token.
[GSI_LOGGER-TOKEN_CLIENT]: The OAuth token was not passed to gapi.client, since the gapi.client library is not loaded in your page.
```

The last line felt a bit weird since I have the following in my `index.html`:
Maybe I'm missing something here?

```
<meta name="google-signin-client_id" content="xxx-xxx.apps.googleusercontent.com">
<script src="https://apis.google.com/js/api.js"></script>
```

Preferably I want to let the user signIn and request scopes once, and the next times the user revisits the application signInSilently will handle most of it in the background.

Sorry if I miss something or misunderstand the way how I should migrate this implementation. I didn't write the original implementation in this application, and I'm not that much of a flutter developer, but I really want to fix this migration for a customer 😀

**Comment by ditman** [comment]
> On page load a signInSilently() got triggered to sign-in the user in the background if he already signed-in manually before. And on the login button the signIn() will get triggered manually. Both of these scenario's used to fire an event in the onCurrentUserChanged listener, but currently only signIn does that.

@ramonvermeulen That is correct. This happened because the former SDK did authentication+authorization in one step, but now they're two separated clients. 

On signInSilently, only authentication happens, so if the plugin gave you an "authenticated but not authorized" user, all your API requests would start failing.

Only when the user is fully authenticated and authorized (after signIn), you get a signed in notification.


```
silentUserLogin() async {
  _googleSignIn.onCurrentUserChanged.listen((GoogleSignInAccount? account) {
    print("user changed"); // never gets called
  });
  await _googleSignIn.signInSilently();
  await _googleSignIn.requestScopes(_googleSignIn.scopes);
}
```

You cannot `signIn` (or `requestScopes`) without user interaction, because those two methods always trigger opening a popup. I think [this previous answer](https://github.com/flutter/flutter/issues/88084#issuecomment-1414278516) may be helpful in your case. TL;DR:

1) Try to `signInSilently` when you build your page. This is probably the code that you already have. That will show the One Tap UX in your page, and if the user agrees, you'll have user authentication information (idToken).
2) You must `signIn` or `requestScopes` in an interactive bit of your app (a button), so users *click* and the browser allows the pop-up to open. Once this is done, you'll get the authorizationToken.

(Your code is probably already doing 1 and 2 above)

> Preferably I want to let the user signIn and request scopes once, and the next times the user revisits the application signInSilently will handle most of it in the background.

That is not going to be possible moving forward, since the new SDK is separating authentication (user ID) from authorization (user permissions). However, if the user has never signed out of your application (or their tokens have expired), their sign-in will be almost automatic (the popup will show up, but will close automatically without users having to do anything extra).

> The last line felt a bit weird since I have the following in my index.html

Don't worry about the last line. That's a trick that the new SDK is attempting to do to be "less breaking" with the users of those libraries. In fact, you should not need any script tag for google authentication to work. Check our super-simple [example app](https://github.com/flutter/plugins/tree/main/packages/google_sign_in/google_sign_in/example).


The biggest issue that you might encounter is that API requests from your app might start failing, and users need to re-authorize access to the `scopes`. The mechanics of authentication should remain the same/similar.

**Comment by ramonvermeulen** [comment]
> > On page load a signInSilently() got triggered to sign-in the user in the background if he already signed-in manually before. And on the login button the signIn() will get triggered manually. Both of these scenario's used to fire an event in the onCurrentUserChanged listener, but currently only signIn does that.
> 
> @ramonvermeulen That is correct. This happened because the former SDK did authentication+authorization in one step, but now they're two separated clients.
> 
> On signInSilently, only authentication happens, so if the plugin gave you an "authenticated but not authorized" user, all your API requests would start failing.
> 
> Only when the user is fully authenticated and authorized (after signIn), you get a signed in notification.
> 
> ```
> silentUserLogin() async {
>   _googleSignIn.onCurrentUserChanged.listen((GoogleSignInAccount? account) {
>     print("user changed"); // never gets called
>   });
>   await _googleSignIn.signInSilently();
>   await _googleSignIn.requestScopes(_googleSignIn.scopes);
> }
> ```
> 
> You cannot `signIn` (or `requestScopes`) without user interaction, because those two methods always trigger opening a popup. I think [this previous answer](https://github.com/flutter/flutter/issues/88084#issuecomment-1414278516) may be helpful in your case. TL;DR:
> 
> 1. Try to `signInSilently` when you build your page. This is probably the code that you already have. That will show the One Tap UX in your page, and if the user agrees, you'll have user authentication information (idToken).
> 2. You must `signIn` or `requestScopes` in an interactive bit of your app (a button), so users _click_ and the browser allows the pop-up to open. Once this is done, you'll get the authorizationToken.
> 
> (Your code is probably already doing 1 and 2 above)
> 
> > Preferably I want to let the user signIn and request scopes once, and the next times the user revisits the application signInSilently will handle most of it in the background.
> 
> That is not going to be possible moving forward, since the new SDK is separating authentication (user ID) from authorization (user permissions). However, if the user has never signed out of your application (or their tokens have expired), their sign-in will be almost automatic (the popup will show up, but will close automatically without users having to do anything extra).
> 
> > The last line felt a bit weird since I have the following in my index.html
> 
> Don't worry about the last line. That's a trick that the new SDK is attempting to do to be "less breaking" with the users of those libraries. In fact, you should not need any script tag for google authentication to work. Check our super-simple [example app](https://github.com/flutter/plugins/tree/main/packages/google_sign_in/google_sign_in/example).
> 
> The biggest issue that you might encounter is that API requests from your app might start failing, and users need to re-authorize access to the `scopes`. The mechanics of authentication should remain the same/similar.

@ditman Thanks a lot for your clarification! With this information I think I need to redesign the flow in the application a little bit to make this work.

In my case `onCurrentUserChanged` still does not get triggered after calling `requestScopes` (manually) when the user is already signed-in with `signInSilently`. I only manage to get the event triggered with the `signIn` function. I changed my login function (triggered on manual click) to the following, and `signInSilently` got called on page load with an event listener on `onCurrentUserChanged`:

```
  Future<void> logIn() async {
    try {
      var signedIn = await _googleSignIn.isSignedIn();
      if (signedIn) {
        await _googleSignIn.requestScopes(_googleSignIn.scopes); // does just scopes
      } else {
        await _googleSignIn.signIn(); // does sign-in + scopes
      }
    } catch (error) {
      throw LogInFailure.fromMessage(error.toString());
    }
  }
```

I can always use `signIn` instead of `requestScopes`, but then the whole `signInSilently` part and one-tap pop-up feels redundant since you have to call `signIn` anyways?

In my opinion it feels really unnatural that after `signInSilently` and the GIS pop-up being showed, the user still has to manually click on a button to trigger another pop-up (e.g. `requestScopes`) to request the authorization scopes. Out of an user perspective if the GIS pop-up is being showed with a chevron the user has the feeling of being signed in. It's quite weird to then have to click on another button to trigger another pop-up to retrieve the right scopes, before being "actually" signed-in into the application. I mean technically you can define the first part as being "signed-in", however in a lot of applications you need additional scopes to be able to actually do something within the application such as calling external API's. In my case it is an application totally build around Google Cloud Search API for which it needs additional scopes.

Out of an implementation perspective I totally get it, they are two different clients one for authentication (GIS one-tap part) and one for authorization (old style pop-up). But for my application it feels more like a step back rather than a step forward.

P.s.

I will clone and test the example app to hopefully get a bit wiser

**Comment by ditman** [comment]
> I can always use signIn instead of requestScopes, but then the whole signInSilently part and one-tap pop-up feels redundant since you have to call signIn anyways?

@ramonvermeulen in your case, maybe it's simpler to just stop doing `signInSilently` on the web, and just use `signIn` when the user clicks the sign in button? This would be the same that your current app is doing in case `silentSignIn` returns a `null` idToken in the current implementation (?).

About `signIn` and `requestScopes`, I don't think you need to over-think this too much. Use `signIn`. If you still use `signInSilently`, and users have consented to it, the popup will go straight to the `requestScopes` part. If the user hasn't completed `signInSilently` (they're not authenticated in the browser, or they aborted the process), `signIn` will sign them in, and then `requestScopes`.

> it feels really unnatural that after signInSilently and the GIS pop-up being showed, the user still has to manually click on a button to trigger another pop-up (e.g. requestScopes) to request the authorization scopes. [...] in a lot of applications you need additional scopes to be able to actually do something within the application such as calling external API's. [...] for my application it feels more like a step back rather than a step forward.

I agree. This plugin is just another user of the new GIS SDK and had very little input on its design. In fact, the way this unfolded has us thinking about [the future of the Google Sign In plugin(s)](https://github.com/flutter/flutter/issues/119300) as a whole.

I can explain why this is moving to a "two step" process, though! In the near future, idTokens may come straight from the user's web browser rather than Google; for example see [FedCM](https://developer.chrome.com/en/docs/privacy-sandbox/fedcm/), and [FIDO2](https://fidoalliance.org/fido2/). In those cases, the Authentication is handled by the browser, **locally**, but the **Authorization** process would still need to call the remote service for permissions.

Anyway, thanks for giving this a shot, I hope you manage to work around the quirks of the new SDKs!

**Comment by ramonvermeulen** [comment]
> > I can always use signIn instead of requestScopes, but then the whole signInSilently part and one-tap pop-up feels redundant since you have to call signIn anyways?
> 
> @ramonvermeulen in your case, maybe it's simpler to just stop doing `signInSilently` on the web, and just use `signIn` when the user clicks the sign in button? Wouldn't this be the same/similar that your current app does when `signInSilently` returns a `null` idToken in the current implementation (?).
> 
> About `signIn` and `requestScopes`, I don't think you need to over-think this too much. Use `signIn`. If you still use `signInSilently`, and users have consented to it, the popup will go straight to the `requestScopes` part. If the user hasn't completed `signInSilently` (they're not authenticated in the browser, or they aborted the process), `signIn` will sign them in, and then `requestScopes`.
> 
> The code of the new plugin does what it does to maximize the compatibility with the previous code, without you having to change much about the sign-in flow (for more details, pls look at the full README!)
> 
> > it feels really unnatural that after signInSilently and the GIS pop-up being showed, the user still has to manually click on a button to trigger another pop-up (e.g. requestScopes) to request the authorization scopes. [...] in a lot of applications you need additional scopes to be able to actually do something within the application such as calling external API's. [...] for my application it feels more like a step back rather than a step forward.
> 
> I agree. This plugin is just another user of the new GIS SDK and had very little input on its design. In fact, the way this unfolded has us thinking about [the future of the Google Sign In plugin(s)](https://github.com/flutter/flutter/issues/119300) as a whole.
> 
> I can explain why this is moving to a "two step" process, though! In the near future, idTokens may come straight from the user's web browser rather than Google; for example see [FedCM](https://developer.chrome.com/en/docs/privacy-sandbox/fedcm/), and [FIDO2](https://fidoalliance.org/fido2/). In those cases, the Authentication is handled by the browser, **locally**, but the **Authorization** process would still need to call the remote service for permissions.
> 
> Anyway, thanks for giving this a shot, I hope you manage to work around the quirks of the new SDKs!

Thanks for the explanation! I eventually got it all working in the application, had to dive a bit into how to persist state with bloc using hydrated bloc and make some structural changes to the application. I save the authHeaders in state now, and log the user out if I get a 403 back from one of the google APIs. The accessToken has a lifetime of 1 hour, so after 1 the user has to manually sign in again. I haven't seen a mechanism to refresh the accessToken yet, or change the lifetime of the accessToken. (Maybe I'm missing something here, if one of these is possible I would be even more happy 😄)

I haven't heard of the **[FedCM](https://developer.chrome.com/en/docs/privacy-sandbox/fedcm/)** and **https://fidoalliance.org/fido2/** concepts before. Thanks for referring to it, makes more sense now why google is separating authentication and authorization to different APIs.

**Comment by ditman** [comment]
> I haven't seen a mechanism to refresh the accessToken yet (can't find a refreshToken in GoogleSignInAuthentication for example), or change the lifetime of the accessToken. (Maybe I'm missing something here, if one of these is possible I would be even more happy 😄 but referring to the [docs](https://cloud.google.com/docs/authentication/token-types#at-lifetime) I guess extending the lifetime is not possible)

@ramonvermeulen This is not possible from the client-side anymore, if you want to extend a users' session, they need to re-authorize the scopes.

The only way that I'm aware of that you get a renew token and all that jazz is if when you use the ["Authorization Code" model](https://developers.google.com/identity/oauth2/web/guides/use-code-model), which then you use **from your server-side** to do requests to Google on your users' behalf, behind the scenes.

Then your users access **your backend** (and not Google's) where you're supposed to have all the data they need.

> I eventually got it all working in the application, had to dive a bit into how to persist state with bloc using hydrated bloc and make some structural changes to the application. I save the authHeaders in state now, and log the user out if I get a 403 back from one of the google APIs.

I think a blog post about this, if you have the energy to write it, would be very beneficial to the community. I can't handle all the integrations with all the possible state management solutions, but we can have links to blog posts describing solutions for the most common architectures!

**Comment by ditman** [comment]
The **PR has been reviewed and labeled as `autosubmit`**, it'll land into the repo as soon as it's ready to accept new changes.

**Comment by ditman** [comment]
Reopening this until we update the core plugin to use the new version!

**Comment by ramonvermeulen** [comment]
> > I haven't seen a mechanism to refresh the accessToken yet (can't find a refreshToken in GoogleSignInAuthentication for example), or change the lifetime of the accessToken. (Maybe I'm missing something here, if one of these is possible I would be even more happy smile but referring to the [docs](https://cloud.google.com/docs/authentication/token-types#at-lifetime) I guess extending the lifetime is not possible)
> 
> @ramonvermeulen This is not possible from the client-side anymore, if you want to extend a users' session, they need to re-authorize the scopes.
> 
> The only way that I'm aware of that you get a renew token and all that jazz is if when you use the ["Authorization Code" model](https://developers.google.com/identity/oauth2/web/guides/use-code-model), which then you use **from your server-side** to do requests to Google on your users' behalf, behind the scenes.
> 
> Then your users access **your backend** (and not Google's) where you're supposed to have all the data they need.
> 
> > I eventually got it all working in the application, had to dive a bit into how to persist state with bloc using hydrated bloc and make some structural changes to the application. I save the authHeaders in state now, and log the user out if I get a 403 back from one of the google APIs.
> 
> I think a blog post about this, if you have the energy to write it, would be very beneficial to the community. I can't handle all the integrations with all the possible state management solutions, but we can have links to blog posts describing solutions for the most common architectures!

Thanks for the information, also nice job on the Migrating to v0.11 section in the readme, explains everything really well in my opinion.

I see, in my case we don't have our own server-side "back-end" for this app. It's just a client calling Google API's for authentication / authorization, and then fetching search results from the Google Cloud Search API (which is basically the back-end). But I don't think it will be that big of a problem that an user has to re-authorize after 1 hour of inactivity, was just wondering if there were other options.

Unfortunately I'm quite packed with work at the moment, so I don't have a lot of time to write a blog post. However, for people reading this and facing the same issues and are already using [**flutter_bloc**](https://pub.dev/packages/flutter_bloc) as state management solution within their application. Be sure to take a look at [**hydrated_bloc**](https://pub.dev/packages/hydrated_bloc) which is the library I used to persist the bloc state, for instance you can store the authHeaders in here so a user can stay "signed-in" with the required scopes for up to one hour.

**Comment by ditman** [comment]
> in my case we don't have our own server-side "back-end" for this app. It's just a client calling Google API's for authentication / authorization, and then fetching search results from the Google Cloud Search API (which is basically the back-end)

@ramonvermeulen yes, this is the case of many people too, you're not alone :)

> I don't think it will be that big of a problem that an user has to re-authorize after 1 hour of inactivity

I *think* users have to re-authorize after 1 hour (even if they're being very busy with your app!)

> I'm quite packed with work at the moment, so I don't have a lot of time to write a blog post.

Again, thanks for giving this a test, checking the documentation and everything! Please, feel no pressure about the blog post **at all**! It's possible that others will fill the gaps. IMO people finding this issue and knowing that their migration "is possible" is quite a good start!

**Comment by TimYusR** [comment]
Hi! @ditman 
I have a problem with the new `google_sign_in` version `6.0.0` on the web.

I call `signIn()` which returns `GoogleSignInAccount`. Then I call `authentication` and hope to get an `idToken` for my backend as it was in the previous version of the package, but the `idToken` is null, instead, I have an `accessToken` which the backend can't exchange on our access token.

Here's my piece of code
```dart
  Future<String> signInWithGoogleIdToken() async {
    GoogleSignInAccount? account;

    try {
      account = await googleSignIn.signIn();
    } catch (e) {
      _catchSignInError(e);
    }

    final authentication = await account?.authentication;

    return authentication?.idToken ?? '';
  }
```

I've read the previous conversation here and think that `idToken` shouldn't be null.
I suppose [this](https://github.com/flutter/flutter/issues/88084#issuecomment-1415084938) is the working code and it's similar to mine (second one).

I will appreciate it if you can give me some advice on what could happen wrong 🙏

**Comment by ditman** [comment]
> I call `signIn()` [...], but the `idToken` is null

@TimYusR, @scrfrk, @NickelGit, @Silipok, @ruslanislan, @JavasquirtDeveloper, @tuanvumaihuynh you're correct. Since `signIn` now uses the Oauth implicit flow to authenticate and authorize users, an `idToken` is not being returned there.

If you need the `idToken`, make sure to call `signInSilently` when you build your page to give an opportunity to the OneTap UX to grab the `idToken` (if the user wants to give it to you).

Here's the relevant information about your use case:

https://pub.dev/packages/google_sign_in_web#idtoken-is-null-in-the-googlesigninaccount-object-after-signin

(Since we have received a bunch of feedback regarding this, I'm going to report the lack of an `idToken` on the Oauth flows as a bug to the GIS team //cc @brdaugherty).

**Comment by TimYusR** [comment]
So I need to call `signInSilently` even if I don't need that. It seems a little bit awkward 😅

@ditman Thanks!
> (Since we have received a bunch of feedback regarding this, I'm going to report the lack of an idToken on the Oauth flows as a bug feature request to the GIS team //cc @brdaugherty -> b/270203049).

**Comment by ditman** [comment]
> It seems a little bit awkward 😅

It is, however it was required to make the new SDK behave similarly to the old one to ease the migration of everybody!

**Comment by pldelattre** [comment]
A few comments after testing: 

1/ using googleSignIn.signIn() works well with firebase auth in the cloud but will break firebase auth emulator since it expects idToken. Error is the following : 

```firebase_auth/the-auth-emulator-only-support-sign-in-with-google.com-using-id-token,-not-access-token.-please-update-your-code-to-use-id-token```

2/ and when using ```signInSilently``` followed by ```signIn``` as in the following code : 

```dart
Future googleLogin() async {
    try {
      await googleSignIn.signInSilently();
      final googleUser = await googleSignIn.signIn();
      if (googleUser == null) return;
      _user = googleUser;

      final googleAuth = await googleUser.authentication;

      final credential = GoogleAuthProvider.credential(
        accessToken: googleAuth.accessToken,
        idToken: googleAuth.idToken,
      );

      await FirebaseAuth.instance.signInWithCredential(credential);
      notifyListeners();
    } catch (e) {
      debugPrint(e.toString());
    }
  }
```
It works well on a chrome session that already knows my google account, however it fails in an incognito window. 
In the incognito window, first click on the login button open the signIn popup dialog (and I suppose it fails to get the idToken because signInSilently failed). However, I would have expected that a retry on the login button would work and instead I get the following error : 

<img width="896" alt="image" src="https://user-images.githubusercontent.com/19731392/221380893-915b0f4a-6d94-4381-a4dd-21166a999ec7.png">

@ditman

**Comment by ditman** [comment]
> "I would have expected that a retry on the login button would work"

@pldelattre It doesn't work, because the retry needs to be done by signInSilently. You should probably refresh the incognito window after sign-in, to let the signInSilently process to work?

**Comment by pldelattre** [comment]
> > "I would have expected that a retry on the login button would work"
> 
> @pldelattre It doesn't work, because the retry needs to be done by signInSilently. You should probably refresh the incognito window after sign-in, to let the signInSilently process to work?

If you look at my googleLogin() function in the message above which is the function I call with the button, you can see that the first line is calling signInSilently and this is why I expected a retry on the button to work. Unfortunately, it does not.

**Comment by ditman** [comment]
> the first line is calling signInSilently and this is why I expected a retry on the button to work. Unfortunately, it does not.

Please, check the JS Console of the browser (in development mode) for logging, it'll let you know if the OneTap UX decided to not show up for any reason. There's a few reasons why the "One Tap UX" may not show, [it's smart](https://developers.google.com/identity/gsi/web/guides/features#exponential_cool-down). Please, use signInSilently as advised: when you build your pages, and not as part as your "signIn" button implementation.

There's also an internal bug that I'm trying to get solved to make this much simpler for everybody; as mentioned [here](https://github.com/flutter/flutter/issues/88084#issuecomment-1439088109).

**Comment by c-seeger** [comment]
With the current solution it feels a bit strange, since google silent login will always be triggered on my LoginPage (using it as stated in the example within the `initState`). My LoginPage contains multiple login options not only google, so google should only be triggered to make any call to google services once the user decides to use it for login but not in advance.

**Comment by ditman** [comment]
> My LoginPage contains multiple login options not only google, so google should only be triggered to make any call to google services once the user decides to use it for login but not in advance.

@c-seeger yes, you're right. I guess your Login page is similar to a "logo tower" with a bunch of login options. In that case, GSI recommends this:

* [Display the Sign In With Google button](https://developers.google.com/identity/gsi/web/guides/display-button)

This is a missing feature of the current `google_sign_in_web`. Could you create a new issue so we can discuss its implementation separately from the more general plugin?

(There's also the problem that users authenticating there would still *not authorize scopes* however!)

**Comment by DanMossa** [comment]
I originally had the code
```dart
  GoogleSignIn googleSignIn = GoogleSignIn(
    clientId: constants.WEB_GOOGLE_CLIENT_ID,
  );

  GoogleSignInAccount? account = await googleSignIn.signIn();
  GoogleSignInAuthentication? auth;
  if (account != null) {
    auth = await account.authentication;
  }

  AuthResponse? signInRes;
  AuthException? error;
  if (auth?.idToken != null) {
    try {
      signInRes = await AuthService.signInWIthOpenID(
        idToken: auth!.idToken!,
        provider: Provider.google,
      );
    } on AuthException catch (e) {
      error = e;
    }
  }
```

And this code gave me the error: `The OAuth token was not passed to gapi.client, since the gapi.client library is not loaded in your page.`

And reading comments it looks like I should be calling `signInSilently` first. Silently signing in works, but then calling `signIn` just hangs.

**Comment by ditman** [comment]
Hello all who must have an `idToken`!

I'm currently implementing the "Google Sign In Button" flow for the web. Please, follow the progress in the corresponding places:

* The issue: https://github.com/flutter/flutter/issues/121781
* The WIP: https://github.com/flutter/packages/pull/3461

The PR adds the ["Google Sign In Button"](https://developers.google.com/identity/gsi/web/guides/display-button) as a Flutter for Web Widget, so Web apps can use that instead of the programmatic `signIn` code.

It also adds a `canAccessScopes` method so the app can check if a logged-in user still has permissions to access all the scopes that are needed.

_Also also:_ updates the `example` app to see all the new changes "in action".

(PS: **I think we should stop updating this issue.** Let's continue the conversation in the issues/PRs that are being actively worked on, or on new ones for anything new that might show up. Thanks for reading!)

**Comment by rekire** [comment]
hey @ramonvermeulen how did you fix your issue with `[GSI_LOGGER-TOKEN_CLIENT]: The OAuth token was not passed to gapi.client, since the gapi.client library is not loaded in your page.`? just using `signIn()` instead of `signInSilently()` doesn't work for me
